### PR TITLE
fix: #520 (moving types out, fixing soundness)

### PR DESCRIPTION
**Pull request by maxkrieger**
# Description

Related issue/PR: #520 

Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context. List any dependencies that are required for this change.

# Implementation strategy and design decisions

Include a high-level summary of the implementation strategy and list important design decisions made, if any.

# Examples with steps to reproduce them

# Checklist

- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] My changes generate no new warnings
- [ ] New and existing tests pass locally using `yarn test`
- [ ] I ran `yarn docs` and there were no errors when generating the HTML site
- [ ] My code follows the style guidelines of this project (e.g.: no ESLint warnings)

# Open questions

Questions that require more discussion or to be addressed in future development:

### CODE REVIEW

**Comment by maxkrieger** [comment]
@wodeni should probs merge asap if it lgty

File: packages/core/src/types/shapeEvalTypes.ts [hunk 1]
```diff
@@ -0,0 +1,477 @@
+import { VarAD } from "./adTypes";
```
Comment (wodeni) [inline packages/core/src/types/shapeEvalTypes.ts:1]: Name change:

File: packages/core/src/types/shapeEvalTypes.ts [hunk 2]
```diff
@@ -0,0 +1,477 @@
...
+  tag: "Done";
+  contents: Value<T>;
+}
+
+export interface IPending<T> {
+  tag: "Pending";
+  contents: Value<T>;
+}
+
+export type Expr =
```
Comment (wodeni) [inline packages/core/src/types/shapeEvalTypes.ts:85]: This belongs to style AST technically

File: packages/core/src/types/shapeEvalTypes.ts [hunk 3]
```diff
@@ -0,0 +1,477 @@
...
+  tag: "StringLit";
+  contents: string;
+}
+
+export interface IBoolLit extends ASTNode {
+  tag: "BoolLit";
+  contents: boolean;
+}
+
+// COMBAK: This seems to be unused, delete
```
Comment (wodeni) [inline packages/core/src/types/shapeEvalTypes.ts:129]: Can remove this and the one below

File: packages/core/src/types/styleASTTypes.ts [hunk 4]
```diff
@@ -0,0 +1,189 @@
+//#region Style AST
+// TODO: unify type name convention (e.g. stop using `I` for interfaces and drop some of the Haskell ported types)
+
+import { ASTNode, Identifier } from "./ASTTypes";
+import { StyVar, BindingForm, Expr, Path } from "./shapeEvalTypes";
```
Comment (wodeni) [inline packages/core/src/types/styleASTTypes.ts:5]: These should all be defined in this file instead

File: packages/core/src/types/substanceASTTypes.ts [hunk 5]
```diff
@@ -0,0 +1,116 @@
+import { LabelMap } from "compiler/Substance";
+import { ASTNode, Identifier } from "./ASTTypes";
+import { TypeConstructor } from "./domainASTTypes";
+import { IStringLit } from "./shapeEvalTypes";
```
Comment (wodeni) [inline packages/core/src/types/substanceASTTypes.ts:4]: Move this to ASTTypes

File: packages/core/src/types/substanceASTTypes.ts [hunk 6]
```diff
@@ -0,0 +1,116 @@
+import { LabelMap } from "compiler/Substance";
```
Comment (wodeni) [inline packages/core/src/types/substanceASTTypes.ts:1]: Move this type to this file

File: packages/core/src/types/types.d.ts [hunk 7]
```diff
@@ -1,1614 +1,2 @@
 // TODO: find a way to put this in index.d.ts
```
Comment (wodeni) [inline packages/core/src/types/types.d.ts:1]: Remove this file after resolving this TODO item?

**Review by wodeni — CHANGES_REQUESTED** [review]
Overall LGTM. Need to move some things around as noted below. Also can you check if you have any circular deps (maybe by https://github.com/benmosher/eslint-plugin-import/blob/master/docs/rules/no-cycle.md?). Circular imports themselves aren't problematic, but it shows illogical module organization quite often.

**Comment by maxkrieger** [comment]
Good point, will cycle check tomorrow

File: packages/core/src/types/shapeEvalTypes.ts (hunk 1 above)
Comment (maxkrieger) [reply packages/core/src/types/shapeEvalTypes.ts:1]: ?

File: packages/core/src/types/shapeEvalTypes.ts (hunk 2 above)
Comment (maxkrieger) [reply packages/core/src/types/shapeEvalTypes.ts:85]: the whole thing? all the sum type members r in there

File: packages/core/src/types/styleASTTypes.ts (hunk 4 above)
Comment (maxkrieger) [reply packages/core/src/types/styleASTTypes.ts:5]: these meaning https://github.com/penrose/penrose/pull/526#discussion_r592051644 ?

File: packages/core/src/types/substanceASTTypes.ts (hunk 5 above)
Comment (maxkrieger) [reply packages/core/src/types/substanceASTTypes.ts:4]: IStringLit?

**Review by wodeni — APPROVED** [review]

**Pull request merged**